### PR TITLE
Refactor/fix Cohere2

### LINKED ISSUE #35547: FA2 broken for Cohere2 if Optional `Mask` is not passed in `forward`
### System Info

transformers==4.48.0.dev0 (from git+https://github.com/huggingface/transformers.git@5615a393691c81e00251e420c73e4d04c6fe22e5)

### Who can help?

@ArthurZucker @Cyrilvallez @SunMarc 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Check our CI test failures:

# Gemma
https://github.com/ModelCloud/GPTQModel/actions/runs/12651906072/job/35253942521#step:12:1164

# Cohere2
https://github.com/ModelCloud/GPTQModel/actions/runs/12651906072/job/35253938235#step:12:922

We enabled FA2 by default on GPTQModel for inference of gptq quantized models and our CI tests are failing for multiple models. This looks like a regression in the `fa2` attention code where `seq_len` is never set if `mask is None`.  FA2 forward requires `seq_len`:

https://github.com/huggingface/transformers/blob/fc74e3976d090b655a8278ca1b70f784ff8b7938/src/transformers/models/cohere2/modeling_cohere2.py#L235-L270

@SunMarc  I don't think this is related to quantization and @ArthurZucker The FA2 code above is broken if `mask` is not passed or `None` as `seq_len` will never be set.  The `mask` param is explicitly declared as `Optional`.

### Expected behavior

Work and not crash.

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

Refactor Cohere2 to meet recent standards, and fix https://github.com/huggingface/transformers/issues/35547 at the same time

### CODE REVIEW

**Comment by Cyrilvallez** [comment]
cc @ArthurZucker, and @alexrs-cohere if you want to have a look!

File: src/transformers/models/cohere2/modeling_cohere2.py [hunk 1]
```diff
@@ -181,117 +175,29 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
 
 
 def eager_attention_forward(
-    config: Cohere2Config,
+    module: nn.Module,
```
Comment (ArthurZucker) [inline src/transformers/models/cohere2/modeling_cohere2.py:178]: cc @ydshieh 😉

File: src/transformers/models/cohere2/modeling_cohere2.py [hunk 2]
```diff
@@ -181,117 +175,29 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
...
-    attn_weights = torch.matmul(query, key_states.transpose(2, 3)) / math.sqrt(config.head_dim)
-
-    if mask is not None:  # no matter the length, we just slice it
-        causal_mask = mask[:, :, :, : key_states.shape[-2]]
+    attention_mask: Optional[torch.Tensor],
+    scaling: float,
+    dropout: float = 0.0,
+    **kwargs,
+):
+    key_states = repeat_kv(key, module.num_key_value_groups)
```
Comment (ArthurZucker) [inline src/transformers/models/cohere2/modeling_cohere2.py:187]: can't we import it?

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/models/cohere2/modeling_cohere2.py (hunk 1 above)
Comment (ydshieh) [reply src/transformers/models/cohere2/modeling_cohere2.py:178]: yes!!!!!!!!!!

File: src/transformers/models/cohere2/modeling_cohere2.py (hunk 2 above)
Comment (Cyrilvallez) [reply src/transformers/models/cohere2/modeling_cohere2.py]: Oh indeed, good point! More code cut 🤗
